### PR TITLE
few additional improvements to recurring plans

### PR DESCRIPTION
**Pull request by PetrDlouhy**
- fix DoesNotExist exception when recurring plan is not present
- add plan_autorenew_at() helper function to provide unified implementation
- add RecurringUserPlan.card_masked_number field

This requires PR #113 to be merged first to preserve order of migrations.

### CODE REVIEW

File: plans/models.py [hunk 1]
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
             return self.expire
         return self.get_plan_extended_from(plan) + timedelta(days=pricing.period)
 
+    def plan_autorenew_at(self):
+        """
+        Helper function which calculates when the plan autorenewal will occur
+        """
+        PLANS_AUTORENEW_BEFORE_DAYS = getattr(settings, 'PLANS_AUTORENEW_BEFORE_DAYS', 0)
```
Comment (Alir3z4) [inline plans/models.py:268]: I think in method scope, variables should not be all capital case ?

File: plans/models.py [hunk 2]
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
...
+        PLANS_AUTORENEW_BEFORE_HOURS = getattr(settings, 'PLANS_AUTORENEW_BEFORE_HOURS', 0)
+        return self.expire - timedelta(days=PLANS_AUTORENEW_BEFORE_DAYS, hours=PLANS_AUTORENEW_BEFORE_HOURS)
+
     def set_plan_renewal(self, order, has_automatic_renewal=True, **kwargs):
         """
         Creates or updates plan renewal information for this userplan with given order
         """
-        self.recurring.delete()
+        if hasattr(self, 'recurring'):
+            self.recurring.delete()
```
Comment (Alir3z4) [inline plans/models.py:277]: I think it would be better to add some logging here. 
Also a single comment would make the operation clear.

For instance why `recurring` object is being deleted and later another one is being created for.

The comment/documentation will make the code much easier to reason about.

File: plans/models.py [hunk 3]
```diff
@@ -261,11 +261,20 @@ def get_plan_extended_until(self, plan, pricing):
             return self.expire
         return self.get_plan_extended_from(plan) + timedelta(days=pricing.period)
 
+    def plan_autorenew_at(self):
```
Comment (Alir3z4) [inline plans/models.py:264]: This method would benefit from a simple unit test.

File: plans/models.py (hunk 2 above)
Comment (Alir3z4) [reply plans/models.py:277]: Also a unit test for this edge case would be a great addition.

**Comment by Alir3z4** [comment]
Great stuff, just did a review and requested some changes.

Thanks for the great work on this.

File: demo/requirements.txt [hunk 4]
```diff
@@ -1,12 +1,14 @@
-django==2.2.13
```
Comment (Alir3z4) [inline demo/requirements.txt:1]: hmm, why django has been removed ?

**Comment by PetrDlouhy** [comment]
@Alir3z4 I have corrected the defects you mentioned and added the unit tests. I also added a few more things:
- fix of the tox settings
- few presets for my development environment
- fix for the `django_next_url_mixin` dependency version

File: setup.cfg [hunk 5]
```diff
@@ -1,4 +1,7 @@
 [tool:pytest]
 python_files=test*.py
 addopts=--tb=short
+[flake8]
+max-line-length = 180
```
Comment (Alir3z4) [inline setup.cfg:5]: I'd suggest 120 will good in case of using type hinting etc, but 180 would be very long and especially for code review that would require horizontal scrolling.

File: demo/requirements.txt (hunk 4 above)
Comment (PetrDlouhy) [reply demo/requirements.txt:1]: Because it collides with Django dependency defined in tox.ini

File: tox.ini [hunk 6]
```diff
@@ -1,6 +1,17 @@
 [tox]
```
Comment (Alir3z4) [inline tox.ini:1]: Additions to this are great, lovely <3

File: demo/requirements.txt (hunk 4 above)
Comment (Alir3z4) [reply demo/requirements.txt:1]: Oh, it's the demo, never mind then, sorry my bad.

File: setup.cfg (hunk 5 above)
Comment (PetrDlouhy) [reply setup.cfg:5]: I can do that, but there are already many lines longer than that in the code. Too many long lines errors in the report was obscuring other more important errors, that's why I added this to the `setup.cfg`

But I can remove this commit from the PR entirely and solve all flake issues in separate PR and add it to the Travis testing, if you are interested.

Comment (Alir3z4) [reply setup.cfg:5]: Hmm, that's unfortunate.
In that case, 180 is OK, the line count can be dealt with in a separate pull request, handling those long lines would pollute the pull request.

Let's carry on then.

**Comment by Alir3z4** [comment]
All good, this looks good to me!

Let's get it in then.
Thanks for the great stuff.
